src/test/java/games/strategy/ui/SwingActionTest.java

```diff
@@ -147,10 +145,61 @@ public void testInvokeAndWaitUninterruptiblyWithRunnable_ShouldRethrowActionUnch
   @Test
   public void testInvokeAndWaitUninterruptiblyWithRunnable_ShouldRethrowActionUncheckedExceptionWhenCalledOnEdt()
       throws Exception {
-    SwingUtilities.invokeAndWait(() -> {
-      assertThrows(IllegalStateException.class,
-          () -> SwingAction.invokeAndWaitUninterruptibly(RUNNABLE_THROWING_EXCEPTION));
+    SwingUtilities.invokeAndWait(() -> assertThrows(
+        IllegalStateException.class,
+        () -> SwingAction.invokeAndWaitUninterruptibly(RUNNABLE_THROWING_EXCEPTION)));
+  }
+
+  @Test
+  public void testInvokeAndWaitUninterruptiblyWithSupplier_ShouldReturnActionResultWhenCalledOffEdt() {
+    assertEquals(Optional.of(VALUE), SwingAction.invokeAndWaitUninterruptibly(() -> VALUE));
+  }
+
+  @Test
+  public void testInvokeAndWaitUninterruptiblyWithSupplier_ShouldReturnActionResultWhenCalledOnEdt() throws Exception {
+    SwingUtilities.invokeAndWait(
+        () -> assertEquals(Optional.of(VALUE), SwingAction.invokeAndWaitUninterruptibly(() -> VALUE)));
+  }
+
+  @Test
+  public void testInvokeAndWaitUninterruptiblyWithSupplier_ShouldReturnEmptyResultWhenInterrupted() throws Exception {
+    final CountDownLatch threadReadyLatch = new CountDownLatch(1);
+    final CountDownLatch testCompleteLatch = new CountDownLatch(1);
+    final AtomicReference<Optional<Object>> result = new AtomicReference<>();
+
+    final Thread thread = new Thread(() -> {
+      result.set(SwingAction.invokeAndWaitUninterruptibly(() -> {
+        threadReadyLatch.countDown();
+        try {
+          testCompleteLatch.await();
+        } catch (final InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
+        return VALUE;
+      }));
     });
+    thread.start();
+    threadReadyLatch.await();
+    thread.interrupt();
```

Comment:
I don't think there's a guarantee `testcompletelatch.await()` is invoked before interrupting and therefore no InterruptedException will be thrown.

Reply:
It should not be necessary for `testCompleteLatch.await()` to be called before interrupting `thread`.  If you look closely, you'll see that we're interrupting `thread`, not the EDT.  `testCompleteLatch.await()` is running on the EDT, not `thread`.  That call is simply there to ensure the function running on the EDT does not exit before we have a chance to interrupt `thread`.

(The `InterruptedException` try-catch block in the EDT is kind of a red herring.  It's there simply because no checked exception can be thrown from the `Supplier`; it has nothing to do with the `InterruptedException` that will be raised on `thread`.  Maybe it would be clearer if I added a `throw new AssertionError("should not happen")` instead of `Thread.currentThread().interrupt()` in the `Supplier`?)

Reply:
Just to ensure there is no race between `thread.interrupt()` and `testCompleteLatch.await()`, I added a `Thread.sleep(5000L)` call immediately before `testCompleteLatch.await()`, and the test continues to pass.

Reply:
Yeah an AssertionError (or just calling fail())  would make everything clearer.
Yes, I missed that we're interrupting the thread and not the EDT

Reply:
e9f43ceb1